File: dsa-learning/dsa-java-code/src/main/java/com/dsa/leetcode/labuladong/stack/P503_NextGreaterElementII.java
package com.dsa.leetcode.labuladong.stack;

import java.util.ArrayDeque;
import java.util.Deque;

public class P503_NextGreaterElementII {

    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        int[] res = new int[n];
        Deque<Integer> stack = new ArrayDeque<>();
        for (int i = 2 * n - 1; i >= 0; i--) {
            while (!stack.isEmpty() && stack.peek() <= nums[i % n]) {
                stack.pop();
            }
            res[i % n] = stack.isEmpty() ? -1 : stack.peek();
            stack.push(nums[i % n]);
        }
        return res;
    }


}

File: dsa-learning/dsa-python-code/com/dsa/leetcode/labuladong/stack/P503_NextGreaterElementII.py
# author: tobebetter9527
# since: 2024/5/7 8:57
from typing import List


class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [-1 for _ in range(n)]
        stack = []
        for i in range(2 * n - 1, -1, -1):
            while stack and stack[-1] <= nums[i % n]:
                stack.pop()
            res[i % n] = stack[-1] if stack else -1
            stack.append(nums[i % n])
        return res
